src/types.rs

```diff
@@ -63,6 +64,29 @@ impl fmt::Display for StateId {
     }
 }
 
+impl FromStr for StateId {
+    type Err = &'static str;
```

Comment:
a bit nicer imo to just use `String` here, and check whitespace

```suggestion
impl FromStr for StateId {
    type Err = String;
    
```